code/modules/experisci/experiment/handlers/experiment_handler.dm

```diff
@@ -0,0 +1,393 @@
+
+
+/**
+ * Hooks on destructive scans to try and run an experiment (When using a handheld handler)
+ */
+/datum/component/experiment_handler/proc/try_run_destructive_experiment(datum/source, list/scanned_atoms)
+	SIGNAL_HANDLER
+	var/atom/movable/our_scanner = parent
+	if (selected_experiment == null)
+		playsound(our_scanner, 'sound/machines/buzz-sigh.ogg', 25)
+		to_chat(our_scanner, "<span>No experiment selected!.</span>")
+		return
+	var/successful_scan
+	for(var/scan_target in scanned_atoms)
+		if(action_experiment(source, scan_target))
+			successful_scan = TRUE
+	if(successful_scan)
+		playsound(our_scanner, 'sound/machines/ping.ogg', 25)
+		to_chat(our_scanner, "<span>The scan succeeds.</span>")
+	else
+		playsound(src, 'sound/machines/buzz-sigh.ogg', 25)
+		our_scanner.say("The scan did not result in anything.")
+/**
+ * Hooks on successful explosions on the doppler array this is attached to
+ */
+/datum/component/experiment_handler/proc/try_run_doppler_experiment(datum/source, turf/epicenter, devastation_range,
+	heavy_impact_range, light_impact_range, took, orig_dev_range, orig_heavy_range, orig_light_range)
+	SIGNAL_HANDLER
+	var/atom/movable/our_array = parent
+	if(action_experiment(source, devastation_range, heavy_impact_range, light_impact_range))
+		playsound(src, 'sound/machines/ping.ogg', 25)
+	else
+		playsound(src, 'sound/machines/buzz-sigh.ogg', 25)
+		our_array.say("Insufficient explosion to contribute to current experiment.")
+
+/**
+ * Announces a message to all experiment handlers
+ *
+ * Arguments:
+ * * message - The message to announce
+ */
+/datum/component/experiment_handler/proc/announce_message_to_all(message)
+	for(var/experiment in GLOB.experiment_handlers)
+		var/datum/component/experiment_handler/experi_handler = experiment
+		var/atom/movable/experi_parent = experi_handler.parent
+		experi_parent.say(message)
+
+/**
+ * Announces a message to this experiment handler
+ *
+ * Arguments:
+ * * message - The message to announce
+ */
+/datum/component/experiment_handler/proc/announce_message(message)
+	var/atom/movable/experi_parent = parent
+	experi_parent.say(message)
+
+/**
+ * Attempts to perform the selected experiment given some arguments
+ */
+/datum/component/experiment_handler/proc/action_experiment(datum/source, ...)
+	// Check if an experiment is selected
+	if (selected_experiment == null && !(config_flags & EXPERIMENT_CONFIG_ALWAYS_ACTIVE))
+		return FALSE
+
+	// Attempt to run
+	var/list/arguments = list(src)
+	arguments = args.len > 1 ? arguments + args.Copy(2) : arguments
+	if (config_flags & EXPERIMENT_CONFIG_ALWAYS_ACTIVE)
+		for (var/datum/experiment/e in linked_web.available_experiments)
+			. = e.actionable(arglist(arguments)) && e.perform_experiment(arglist(arguments))
+	else
+		// Returns true if the experiment was succesfuly handled
+		. = selected_experiment.actionable(arglist(arguments)) && selected_experiment.perform_experiment(arglist(arguments))
```

Comment:
```suggestion
		return selected_experiment.actionable(arglist(arguments)) && selected_experiment.perform_experiment(arglist(arguments))
```

...though with the behavior of the other branch, it feels to me this should just be condensed down to:

````dm
var/datum/experiment/experiment = (config_flags & EXPERIMENT_CONFIG_ALWAYS_ACTIVE) ? linked_web.available_experiments[1] : selected_experiment
experiment.actionable(arglist(arguments)) && selected_experiment.perform_experiment(arglist(arguments))